### PR TITLE
Wavelet packets: extend to nD and support subsets of the axes

### PR DESCRIPTION
**Pull request by grlee77**
This PR is built on top of #392, so the first 4 commits here overlap with that one.

This PR adds two new features to wavelet packets
   - n-dimensional support via NodeND and WaveletPacketND
   - allow specifying the axes to transform as in other DWT-based functions

**AXIS-SPECIFIC TRANSFORM DETAILS**
The axis-specific support is straightforward. It just involves using the existing `axis`/`axes` argument in the underlying DWT routines.  Basically, `WaveletPacket` can take an `axis` argument and `WaveletPacket2D`/`WaveletPacketND` can take an `axes` argument to specify which axes to transform.  The default axes are the same as in the corresponding DWT functions.

**N-DIMENSIONAL SUPPORT DETAILS**
To make the conventions for the n-dimensional case clearer, let's review the 1D and 2D behavior:

Assume a 1D a wavelet packet was generated via something like:
```python
import pywt
ecg = pywt.data.ecg()
wp = pywt.WaveletPacket(ecg, 'sym5', maxlevel=4)
```
To access a second level node corresponding to the lowpass ('a') followed by highpass ('d') filtering, one would index by concatenating:
```python
wp['ad']
```

2D works in the same way except instead of 'a' and 'd', there are four subbands named 'a', 'h', 'v', and 'd'. 

However, this sort of single-character naming convention does not extend well to the general n-dimensional case.  

```python
import numpy as np
import pywt
wp_nd = pywt.WaveletPacketND(np.random.randn(32, 32, 32), 'sym4', maxlevel=3)
```
In this PR, at each level, the nodes have the same names as the coefficients produced by `dwtn` or `wavedecn` (e.g. 'aad' in 3D indicates lowpass along the first two axes and highpass along the 3rd). The node corresponding to 'aad' followed by 'daa' would then be indexed by:

```python
wp['aaddaa']
```
However from the string `aaddaa` alone it is hard to know if that refers to the 6th level of a 1D decomposition, 3rd-level of a 2D transform, 2nd-level of a 3D transform or the first level of a 6D transform.  To resolve this ambiguity, I modified the `BaseNode` class to also accept a tuple of strings such as:
```python
wp[('aad', 'daa')]
```
In this form it is easy to see that we are indexing into the 2nd level of a 3D wavelet packet transform.

This tuple-based indexing was implemented in BaseNode, so a side effect of this PR is that the 1D and 2D cases can also now use this type of indexing.  (e.g. `wp[('a', 'd')]` for the 1D example above)

The tests in `wp_nd.py` were created by duplicating the 2D test cases and changing node names to match the naming conventions used by WaveletPacketND.

### CODE REVIEW

**Comment by rgommers** [comment]
Wow, that's a big one. Looks like a good idea. Leave till after 1.0 I would say?

**Comment by grlee77** [comment]
>  Leave till after 1.0 I would say?

Yeah, that is fine. I also have a branch where I implemented a couple of best basis or near best basis algorithms. I think the wavelet packets are more useful once one has a means of choosing between all of the potential bases. The best basis approach was covered under a US patent, but it has expired at this point (the patent is from 1991).

**Comment by grlee77** [comment]
Also, a couple of the files are a near copy of the 2D case, just using the nD syntax to access the coefficients. This includes the demo and `test_wpnd.py`.  It is mainly just the NodeND and WaveletPacketND classes that need to be closely reviewed.

**Comment by rgommers** [comment]
Ouch, I never got around to reviewing this. Let's mark for 1.2, this does look like a very useful enhancement.

File: demo/wp_nd.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+plt.imshow(arr, interpolation="nearest", cmap=plt.cm.gray)
+
+fig = plt.figure()
+i = 1
+nsubplots = len(wp2.get_level(maxlevel, 'natural'))
+nrows = int(np.floor(np.sqrt(nsubplots)))
+ncols = int(np.ceil(nsubplots/nrows))
+for node in wp2.get_level(maxlevel, 'natural'):
+    ax = fig.add_subplot(nrows, ncols, i)
+    ax.set_title("%s" % (node.pretty_path, ))
```
Comment (rgommers) [inline demo/wp_nd.py:26]: `pretty_path` is not an attribute. should it just be `path`? Or `path_tuple`?

File: demo/wp_nd.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    ax.imshow(np.sqrt(np.abs(node.data)), origin='image',
```
Comment (rgommers) [inline demo/wp_nd.py:27]: `ValueError: 'image' is not a valid value for origin; supported values are 'upper', 'lower'`

File: doc/source/ref/wavelet-packets.rst [hunk 3]
```diff
@@ -326,3 +373,76 @@ WaveletPacket2D and WaveletPacket2D tree Node2D
...
+
+  .. method:: reconstruct([update=True])
+
+     Reconstruct data from the subnodes.
+
+     :param update: A boolean indicating whether the coefficients of the
+                    current node and its subnodes will be replaced with values
+                    from the reconstruction.
+
+WaveletPacketND and WaveletPacketND tree NodeND
```
Comment (rgommers) [inline doc/source/ref/wavelet-packets.rst:385]: `WaveletPacketND tree NodeND` looks a little odd.

**Review by rgommers — CHANGES_REQUESTED** [review]
The demo seems broken. Other than that it looks good. There's a lot of code here, I can't review it all in detail but can just play around with it and am happy to see this merged if that doesn't turn up anything unexpected.

File: demo/wp_nd.py (hunk 2 above)
Comment (rgommers) [reply demo/wp_nd.py]: fixed, this was an MPL issue.

File: demo/wp_nd.py (hunk 1 above)
Comment (rgommers) [reply demo/wp_nd.py]: I ran the demo, this was meant to be `path_tuple` I believe - made that fix.

Comment (grlee77) [reply demo/wp_nd.py]: thanks, that is probably correct. I will take a quick look again.

**Comment by grlee77** [comment]
I was playing around with this nD support a long time ago, but I don't think 3D (or nD) wavelet packets are commonly used. Since we provide nD for everything else, though, it seems consistent to offer it.

File: doc/source/ref/wavelet-packets.rst (hunk 3 above)
Comment (grlee77) [reply doc/source/ref/wavelet-packets.rst]: I just changed it to `WaveletPacketND and NodeND` here (as well as for the 1D and 2D cases which used the same wording).

**Comment by rgommers** [comment]
> I was playing around with this nD support a long time ago, but I don't think 3D (or nD) wavelet packets are commonly used. Since we provide nD for everything else, though, it seems consistent to offer it.

That does make sense to me.

**Review by rgommers — APPROVED** [review]
I had another look through this, it all looks good and seems to work. Let's give this a go, thanks @grlee77!

**Pull request merged**